// @covalenthq/client-sdk/dist/services/CovalentClient.js
import { SecurityService } from "./SecurityService";
import { BalanceService } from "./BalanceService";
import { BaseService } from "./BaseService";
import { NftService } from "./NftService";
import { PricingService } from "./PricingService";
import { TransactionService } from "./TransactionService";
import { XykService } from "./XykService";
import { ApiKeyValidator } from "../util/ApiKeyValidator";
export const userAgent = "com.covalenthq.sdk.typescript/1.0.2";
export class Response {
}
/**
 * CovalentClient Class
 */
export class CovalentClient {
    constructor(apiKey, settings, source) {
        const { debug = false, threadCount = 3, enableRetry = true } = settings || {};
        const validator = new ApiKeyValidator(apiKey);
        this._is_key_valid = validator.isValidApiKey();
        this.SecurityService = new SecurityService(apiKey, debug, threadCount, this._is_key_valid, enableRetry, source);
        this.BalanceService = new BalanceService(apiKey, debug, threadCount, this._is_key_valid, enableRetry, source);
        this.BaseService = new BaseService(apiKey, debug, threadCount, this._is_key_valid, enableRetry, source);
        this.NftService = new NftService(apiKey, debug, threadCount, this._is_key_valid, enableRetry, source);
        this.PricingService = new PricingService(apiKey, debug, threadCount, this._is_key_valid, enableRetry, source);
        this.TransactionService = new TransactionService(apiKey, debug, threadCount, this._is_key_valid, enableRetry, source);
        this.XykService = new XykService(apiKey, debug, threadCount, this._is_key_valid, enableRetry, source);
    }
}
/**
 * @deprecated
 * Please use {@link CovalentClient} going forward! This method will be removed after Oct 31, 2023.
 * @see #CovalentClient
 * CovalentClient Class
 */
export class Client {
    constructor(apiKey, settings, source) {
        const { debug = false, threadCount = 3, enableRetry = true } = settings || {};
        const validator = new ApiKeyValidator(apiKey);
        this._is_key_valid = validator.isValidApiKey();
        this.SecurityService = new SecurityService(apiKey, debug, threadCount, this._is_key_valid, enableRetry, source);
        this.BalanceService = new BalanceService(apiKey, debug, threadCount, this._is_key_valid, enableRetry, source);
        this.BaseService = new BaseService(apiKey, debug, threadCount, this._is_key_valid, enableRetry, source);
        this.NftService = new NftService(apiKey, debug, threadCount, this._is_key_valid, enableRetry, source);
        this.PricingService = new PricingService(apiKey, debug, threadCount, this._is_key_valid, enableRetry, source);
        this.TransactionService = new TransactionService(apiKey, debug, threadCount, this._is_key_valid, enableRetry, source);
        this.XykService = new XykService(apiKey, debug, threadCount, this._is_key_valid, enableRetry, source);
    }
}
