// @covalenthq/client-sdk/dist/util/Chains.js
export var Chains;
(function (Chains) {
    Chains["BTC_MAINNET"] = "btc-mainnet";
    Chains["ETH_MAINNET"] = "eth-mainnet";
    Chains["MATIC_MAINNET"] = "matic-mainnet";
    Chains["BSC_MAINNET"] = "bsc-mainnet";
    Chains["AVALANCHE_MAINNET"] = "avalanche-mainnet";
    Chains["OPTIMISM_MAINNET"] = "optimism-mainnet";
    Chains["FANTOM_MAINNET"] = "fantom-mainnet";
    Chains["MOONBEAM_MAINNET"] = "moonbeam-mainnet";
    Chains["MOONBEAM_MOONRIVER"] = "moonbeam-moonriver";
    Chains["RSK_MAINNET"] = "rsk-mainnet";
    Chains["ARBITRUM_MAINNET"] = "arbitrum-mainnet";
    Chains["PALM_MAINNET"] = "palm-mainnet";
    Chains["KLAYTN_MAINNET"] = "klaytn-mainnet";
    Chains["HECO_MAINNET"] = "heco-mainnet";
    Chains["NERVOS_GODWOKEN_MAINNET"] = "nervos-godwoken-mainnet";
    Chains["AXIE_MAINNET"] = "axie-mainnet";
    Chains["EVMOS_MAINNET"] = "evmos-mainnet";
    Chains["ASTAR_MAINNET"] = "astar-mainnet";
    Chains["IOTEX_MAINNET"] = "iotex-mainnet";
    Chains["HARMONY_MAINNET"] = "harmony-mainnet";
    Chains["CRONOS_MAINNET"] = "cronos-mainnet";
    Chains["AURORA_MAINNET"] = "aurora-mainnet";
    Chains["EMERALD_PARATIME_MAINNET"] = "emerald-paratime-mainnet";
    Chains["BOBA_MAINNET"] = "boba-mainnet";
    Chains["ETH_GOERLI"] = "eth-goerli";
    Chains["MATIC_MUMBAI"] = "matic-mumbai";
    Chains["AVALANCHE_TESTNET"] = "avalanche-testnet";
    Chains["BSC_TESTNET"] = "bsc-testnet";
    Chains["MOONBEAM_MOONBASE_ALPHA"] = "moonbeam-moonbase-alpha";
    Chains["RSK_TESTNET"] = "rsk-testnet";
    Chains["ARBITRUM_GOERLI"] = "arbitrum-goerli";
    Chains["FANTOM_TESTNET"] = "fantom-testnet";
    Chains["PALM_TESTNET"] = "palm-testnet";
    Chains["HECO_TESTNET"] = "heco-testnet";
    Chains["NERVOS_GODWOKEN_TESTNET"] = "nervos-godwoken-testnet";
    Chains["EVMOS_TESTNET"] = "evmos-testnet";
    Chains["IOTEX_TESTNET"] = "iotex-testnet";
    Chains["HARMONY_TESTNET"] = "harmony-testnet";
    Chains["AURORA_TESTNET"] = "aurora-testnet";
    Chains["SCROLL_L2_TESTNET"] = "scroll-l2-testnet";
    Chains["SCROLL_SEPOLIA_TESTNET"] = "scroll-sepolia-testnet";
    Chains["COVALENT_INTERNAL_NETWORK_V1"] = "covalent-internal-network-v1";
    Chains["DEFI_KINGDOMS_MAINNET"] = "defi-kingdoms-mainnet";
    Chains["SWIMMER_MAINNET"] = "swimmer-mainnet";
    Chains["BOBA_AVALANCHE_MAINNET"] = "boba-avalanche-mainnet";
    Chains["BOBA_BOBABEAM_MAINNET"] = "boba-bobabeam-mainnet";
    Chains["BOBA_BNB_MAINNET"] = "boba-bnb-mainnet";
    Chains["BOBA_RINKEBY_TESTNET"] = "boba-rinkeby-testnet";
    Chains["BOBA_BOBABASE_TESTNET"] = "boba-bobabase-testnet";
    Chains["BOBA_BNB_TESTNET"] = "boba-bnb-testnet";
    Chains["BOBA_AVALANCHE_TESTNET"] = "boba-avalanche-testnet";
    Chains["KLAYTN_TESTNET"] = "klaytn-testnet";
    Chains["GATHER_MAINNET"] = "gather-mainnet";
    Chains["GATHER_TESTNET"] = "gather-testnet";
    Chains["SKALE_CALYPSO"] = "skale-calypso";
    Chains["SKALE_MAINNET"] = "skale-mainnet";
    Chains["SKALE_RAZOR"] = "skale-razor";
    Chains["AVALANCHE_DEXALOT_MAINNET"] = "avalanche-dexalot-mainnet";
    Chains["SKALE_OMNUS"] = "skale-omnus";
    Chains["AVALANCHE_DEXALOT_TESTNET"] = "avalanche-dexalot-testnet";
    Chains["ASTAR_SHIBUYA"] = "astar-shibuya";
    Chains["CRONOS_TESTNET"] = "cronos-testnet";
    Chains["DEFI_KINGDOMS_TESTNET"] = "defi-kingdoms-testnet";
    Chains["METIS_MAINNET"] = "metis-mainnet";
    Chains["METIS_STARDUST"] = "metis-stardust";
    Chains["MILKOMEDA_A1_MAINNET"] = "milkomeda-a1-mainnet";
    Chains["MILKOMEDA_A1_DEVNET"] = "milkomeda-a1-devnet";
    Chains["MILKOMEDA_C1_MAINNET"] = "milkomeda-c1-mainnet";
    Chains["MILKOMEDA_C1_DEVNET"] = "milkomeda-c1-devnet";
    Chains["SWIMMER_TESTNET"] = "swimmer-testnet";
    Chains["SOLANA_MAINNET"] = "solana-mainnet";
    Chains["SKALE_EUROPA"] = "skale-europa";
    Chains["METER_MAINNET"] = "meter-mainnet";
    Chains["METER_TESTNET"] = "meter-testnet";
    Chains["SKALE_EXORDE"] = "skale-exorde";
    Chains["BOBA_GOERLI"] = "boba-goerli";
    Chains["NEON_TESTNET"] = "neon-testnet";
    Chains["SKALE_STAGING_UUM"] = "skale-staging-uum";
    Chains["SKALE_STAGING_LCC"] = "skale-staging-lcc";
    Chains["ARBITRUM_NOVA_MAINNET"] = "arbitrum-nova-mainnet";
    Chains["CANTO_MAINNET"] = "canto-mainnet";
    Chains["BITTORRENT_MAINNET"] = "bittorrent-mainnet";
    Chains["BITTORRENT_TESTNET"] = "bittorrent-testnet";
    Chains["FLARENETWORKS_FLARE_MAINNET"] = "flarenetworks-flare-mainnet";
    Chains["FLARENETWORKS_FLARE_TESTNET"] = "flarenetworks-flare-testnet";
    Chains["FLARENETWORKS_CANARY_MAINNET"] = "flarenetworks-canary-mainnet";
    Chains["FLARENETWORKS_CANARY_TESTNET"] = "flarenetworks-canary-testnet";
    Chains["KCC_MAINNET"] = "kcc-mainnet";
    Chains["KCC_TESTNET"] = "kcc-testnet";
    Chains["POLYGON_ZKEVM_TESTNET"] = "polygon-zkevm-testnet";
    Chains["LINEA_TESTNET"] = "linea-testnet";
    Chains["BASE_TESTNET"] = "base-testnet";
    Chains["MANTLE_TESTNET"] = "mantle-testnet";
    Chains["SCROLL_ALPHA_TESTNET"] = "scroll-alpha-testnet";
    Chains["OASYS_MAINNET"] = "oasys-mainnet";
    Chains["OASYS_TESTNET"] = "oasys-testnet";
    Chains["FINDORA_MAINNET"] = "findora-mainnet";
    Chains["FINDORA_FORGE_TESTNET"] = "findora-forge-testnet";
    Chains["SX_MAINNET"] = "sx-mainnet";
    Chains["OASIS_SAPPHIRE_MAINNET"] = "oasis-sapphire-mainnet";
    Chains["OASIS_SAPPHIRE_TESTNET"] = "oasis-sapphire-testnet";
    Chains["OPTIMISM_GOERLI"] = "optimism-goerli";
    Chains["POLYGON_ZKEVM_MAINNET"] = "polygon-zkevm-mainnet";
    Chains["HORIZEN_YUMA_TESTNET"] = "horizen-yuma-testnet";
    Chains["CLV_PARACHAIN"] = "clv-parachain";
    Chains["ENERGI_MAINNET"] = "energi-mainnet";
    Chains["ENERGI_TESTNET"] = "energi-testnet";
    Chains["HORIZEN_GOBI_TESTNET"] = "horizen-gobi-testnet";
    Chains["ETH_SEPOLIA"] = "eth-sepolia";
    Chains["SKALE_NEBULA"] = "skale-nebula";
    Chains["SKALE_BATTLEGROUND"] = "skale-battleground";
    Chains["AVALANCHE_MELD_TESTNET"] = "avalanche-meld-testnet";
    Chains["GUNZILLA_TESTNET"] = "gunzilla-testnet";
    Chains["ULTRON_MAINNET"] = "ultron-mainnet";
    Chains["ULTRON_TESTNET"] = "ultron-testnet";
    Chains["ZORA_MAINNET"] = "zora-mainnet";
    Chains["ZORA_GOERLI_TESTNET"] = "zora-goerli-testnet";
    Chains["NEON_MAINNET"] = "neon-mainnet";
    Chains["AVALANCHE_SHRAPNEL_MAINNET"] = "avalanche-shrapnel-mainnet";
    Chains["BASE_MAINNET"] = "base-mainnet";
    Chains["MANTLE_MAINNET"] = "mantle-mainnet";
    Chains["AVALANCHE_LOCO_LEGENDS_MAINNET"] = "avalanche-loco-legends-mainnet";
    Chains["LINEA_MAINNET"] = "linea-mainnet";
    Chains["HORIZEN_EON_MAINNET"] = "horizen-eon-mainnet";
    Chains["AVALANCHE_NUMBERS"] = "avalanche-numbers";
    Chains["AVALANCHE_DOS"] = "avalanche-dos";
    Chains["AVALANCHE_STEP_NETWORK"] = "avalanche-step-network";
    Chains["AVALANCHE_XPLUS"] = "avalanche-xplus";
    Chains["AVALANCHE_XANACHAIN"] = "avalanche-xanachain";
    Chains["AVALANCHE_MELD_MAINNET"] = "avalanche-meld-mainnet";
    Chains["OPSIDE_PUBLIC_ZKEVM"] = "opside-public-zkevm";
    Chains["OPSIDE_LAW_CHAIN"] = "opside-law-chain";
    Chains["AVALANCHE_SHRAPNEL_TESTNET"] = "avalanche-shrapnel-testnet";
    Chains["AVALANCHE_LOCO_LEGENDS_TESTNET"] = "avalanche-loco-legends-testnet";
    Chains["OPSIDE_CB_ZKEVM"] = "opside-cb-zkevm";
    Chains["OPSIDE_PRE_ALPHA_TESTNET"] = "opside-pre-alpha-testnet";
    Chains["OPSIDE_ERA7"] = "opside-era7";
    Chains["OPSIDE_XTHRILL"] = "opside-xthrill";
    Chains["ZKSYNC_MAINNET"] = "zksync-mainnet";
    Chains["METIS_TESTNET"] = "metis-testnet";
    Chains["ZKSYNC_TESTNET"] = "zksync-testnet";
    Chains["AVALANCHE_BLITZ_TESTNET"] = "avalanche-blitz-testnet";
    Chains["AVALANCHE_D_CHAIN_TESTNET"] = "avalanche-d-chain-testnet";
    Chains["AVALANCHE_GREEN_DOT_TESTNET"] = "avalanche-green-dot-testnet";
    Chains["AVALANCHE_MINTARA_TESTNET"] = "avalanche-mintara-testnet";
    Chains["AVALANCHE_BEAM_TESTNET"] = "avalanche-beam-testnet";
    Chains["BNB_META_APES_MAINNET"] = "bnb-meta-apes-mainnet";
    Chains["BNB_ANTIMATTER_MAINNET"] = "bnb-antimatter-mainnet";
    Chains["BNB_ANTIMATTER_TESTNET"] = "bnb-antimatter-testnet";
    Chains["BNB_OPBNB_TESTNET"] = "bnb-opbnb-testnet";
    Chains["OPSIDE_DEBOX"] = "opside-debox";
    Chains["OPSIDE_JACKBOT"] = "opside-jackbot";
    Chains["OPSIDE_ODX_ZKEVM_TESTNET"] = "opside-odx-zkevm-testnet";
    Chains["OPSIDE_READON_CONTENT_TESTNET"] = "opside-readon-content-testnet";
    Chains["OPSIDE_RELATION"] = "opside-relation";
    Chains["OPSIDE_SOQUEST_ZKEVM"] = "opside-soquest-zkevm";
    Chains["OPSIDE_VIP3"] = "opside-vip3";
    Chains["OPSIDE_ZKMETA"] = "opside-zkmeta";
    Chains["AVALANCHE_PULSAR_TESTNET"] = "avalanche-pulsar-testnet";
    Chains["AVALANCHE_UPTN"] = "avalanche-uptn";
    Chains["BNB_FNCY_MAINNET"] = "bnb-fncy-mainnet";
    Chains["ZETACHAIN_TESTNET"] = "zetachain-testnet";
    Chains["KINTO_TESTNET"] = "kinto-testnet";
    Chains["MODE_TESTNET"] = "mode-testnet";
    Chains["LOOT_MAINNET"] = "loot-mainnet";
    Chains["BNB_FNCY_TESTNET"] = "bnb-fncy-testnet";
    Chains["MANTA_TESTNET"] = "manta-testnet";
    Chains["PGN_MAINNET"] = "pgn-mainnet";
    Chains["PGN_TESTNET"] = "pgn-testnet";
    Chains["GNOSIS_MAINNET"] = "gnosis-mainnet";
    Chains["GNOSIS_TESTNET"] = "gnosis-testnet";
    Chains["ROLLUX_MAINNET"] = "rollux-mainnet";
    Chains["ROLLUX_TESTNET"] = "rollux-testnet";
    Chains["TAIKO_JOLNIR_TESTNET"] = "taiko-jolnir-testnet";
    Chains["OPTIMISM_SEPOLIA"] = "optimism-sepolia";
    Chains["BNB_OPBNB_MAINNET"] = "bnb-opbnb-mainnet";
    Chains["TELOS_MAINNET"] = "telos-mainnet";
    Chains["TELOS_TESTNET"] = "telos-testnet";
    Chains["AVALANCHE_HUBBLE_EXCHANGE_TESTNET"] = "avalanche-hubble-exchange-testnet";
    Chains["AVALANCHE_MIHO_TESTNET"] = "avalanche-miho-testnet";
    Chains["AVALANCHE_BULLETIN_TESTNET"] = "avalanche-bulletin-testnet";
    Chains["AVALANCHE_KIWI_TESTNET"] = "avalanche-kiwi-testnet";
    Chains["AVALANCHE_HERO_TESTNET"] = "avalanche-hero-testnet";
    Chains["AVALANCHE_AVACLOUD_TESTNET"] = "avalanche-avacloud-testnet";
    Chains["AVALANCHE_THIRDWEB_TESTNET"] = "avalanche-thirdweb-testnet";
    Chains["AVALANCHE_MONDRIAN_TESTNET"] = "avalanche-mondrian-testnet";
    Chains["AVALANCHE_CONDUIT_TESTNET"] = "avalanche-conduit-testnet";
    Chains["AVALANCHE_NMAC_TESTNET"] = "avalanche-nmac-testnet";
    Chains["AVALANCHE_ORDERLY_TESTNET"] = "avalanche-orderly-testnet";
    Chains["AVALANCHE_AMPLIFY_TESTNET"] = "avalanche-amplify-testnet";
    Chains["AVALANCHE_MIRAI_TESTNET"] = "avalanche-mirai-testnet";
    Chains["AVALANCHE_WAGMI_TESTNET"] = "avalanche-wagmi-testnet";
    Chains["AVALANCHE_PLAYA3ULL_TESTNET"] = "avalanche-playa3ull-testnet";
    Chains["AVALANCHE_BEAM_MAINNET"] = "avalanche-beam-mainnet";
    Chains["SCROLL_MAINNET"] = "scroll-mainnet";
    Chains["ETH_HOLESKY"] = "eth-holesky";
    Chains["TOMOCHAIN_MAINNET"] = "tomochain-mainnet";
    Chains["TOMOCHAIN_TESTNET"] = "tomochain-testnet";
    Chains["AVALANCHE_JONO11_TESTNET"] = "avalanche-jono11-testnet";
    Chains["BASE_SEPOLIA_TESTNET"] = "base-sepolia-testnet";
    Chains["XAI_TESTNET"] = "xai-testnet";
    Chains["ARBITRUM_SEPOLIA"] = "arbitrum-sepolia";
    Chains["LUMOZ_PUBLIC_ZKSYNC_V2"] = "lumoz-public-zksync-v2";
    Chains["LUMOZ_DECIBLING"] = "lumoz-decibling";
    Chains["LUMOZ_STARK_SPORT"] = "lumoz-stark-sport";
    Chains["AVALANCHE_LT0_TESTNET"] = "avalanche-lt0-testnet";
    Chains["AVALANCHE_LT1_TESTNET"] = "avalanche-lt1-testnet";
    Chains["AVALANCHE_LT2_TESTNET"] = "avalanche-lt2-testnet";
    Chains["AVALANCHE_LT3_TESTNET"] = "avalanche-lt3-testnet";
    Chains["AVALANCHE_LT4_TESTNET"] = "avalanche-lt4-testnet";
    Chains["AVALANCHE_LT5_TESTNET"] = "avalanche-lt5-testnet";
    Chains["SYNDR_TESTNET"] = "syndr-testnet";
    Chains["CROSSFI_EVM_TESTNET"] = "crossfi-evm-testnet";
    Chains["CELO_MAINNET"] = "celo-mainnet";
    Chains["TAIKO_KATLA_TESTNET"] = "taiko-katla-testnet";
    Chains["MOVEMENT_MEVM_TESTNET"] = "movement-mevm-testnet";
    Chains["ZORA_SEPOLIA_TESTNET"] = "zora-sepolia-testnet";
    Chains["MERLIN_MAINNET"] = "merlin-mainnet";
    Chains["MERLIN_TESTNET"] = "merlin-testnet";
    Chains["AVALANCHE_HUBBLE_EXCHANGE_MAINNET"] = "avalanche-hubble-exchange-mainnet";
    Chains["XAI_MAINNET"] = "xai-mainnet";
})(Chains || (Chains = {}));
